Severe logs without error data #4550

wkurniawan07 opened the pull request:
Fixes #4550
Here's what happened: take a look at these lines in `ControllerServlet`:
```java
MimeMessage email = new Logic().emailErrorReport(req, e); // can be null
log.severe(ActivityLogEntry.generateSystemErrorReportLogMessage(req, email));
```
Here, if the variable `email` is null, the log will have no data, which is what is seen in the issue thread.
How can the variable `email` be null? Further probing will bring us to this function in `Emails.java`:
```java
public MimeMessage sendErrorReport(HttpServletRequest req, Throwable error) {
    MimeMessage email = null;
    try {
        ...
        email = generateSystemErrorEmail(...) // this function can throw Exception
        ...
    } catch (Exception e) {
        logSevereForErrorInSendingItem("crash report", email, e);
    }
    return email;
}
```
That is, the `email` will be null if there is an error in generating the system error email.

To alleviate, when such failure occurs, an alternative severe log will be done from `Emails.java` which will show the error's stack trace, on top of also sending the reason for why the crash report failed to send (i.e if the crash report failed to send the admin will receive _two_ severe logs: one logging the error that's supposedly reported in the crash report, and another one logging why the crash report failed to send).

(P.S the newly added stack traces will also have no line breaks)

thyageshm commented:
Hi I think there is a more fundamental problem over here: when there is an error in sending email, we have to detect it and log the message body as a severe log in this format:
1. Mention that the error was detected in sending an email (which you have done by specifying "crash report" as an argument to `logSevereForErrorInSendingItem`)
2. specify what the error was while trying to send (which is also done in `logSevereForErrorInSendingItem`)
3. include the ORIGINAL message that was tried to be sent (this is not done)

Example of a case where this problem can occur (and is not covered): When trying to compile the severe logs and sending an email (in `CompileLogsServlet.java` and `CompileLogs.java`). They try to compile the logs for the last our and send an email. It is still possible for this email to have an error. When such an error occurs, once again the original message (which is compiled logs in this case) are "lost".

wkurniawan07 commented:
I don't agree with 3. The very reason why the log can be empty is when the email generation fails. In this case, there's no such thing as "the original message"; there's not even a "message" to begin with, since it is not successfully generated. Since the only important part of the message is the error stack trace, I added the alternative severe logging for the lost "message", which solves the problem.

`CompileLogs` is a different problem (the issue thread points to `ControllerServlet` which I have fixed), although I could fix it here as well if needed.

thyageshm commented:
Thats correct, I meant the error message (since there is no email to talk about...)

I think the stack trace the prof gave was an example. The real issue is that the stack traces in emails have empty messages.

thyageshm commented:
A more specific way to handle this issue would be catch the email creation specific exceptions and log them separately. Also, for the case of CompileLogs, if you think it should be done in another issue, I really dont mind. Especially since the solution for that problem is slightly more involved and requires a little more digging

wkurniawan07 commented:
That issue is already resolved. See the changes: only when the email is non-null then there will be a severe logging from `ControllerServlet` side. Otherwise, the severe logging is done from `Emails` side. No message/stack trace is lost in this way.

thyageshm commented:
Yes, you have solved what the issue required. On further investigation there is a similar issue which has led me to believe that we have to solve this issue by fundamentally handling "crash report" error differently: that is we have to include the error message that we wanted to send in that report as part of the log. So, logging the error message separately as you have done in lines 808, 809 in Emails.java is not a scalable solution (that is, everywhere I have an email error, I have to log the original error message separately.)

Rather, I would prefer having a separate method specifically to handle such email errors

wkurniawan07 commented:
I could probe more into that, but I'll need more time then. I might be able to solve the `CompileLogs` issue concurrently. But then I'll have to take this off the next milestone.

thyageshm commented:
Keep this one in this milestone and open another issue

wkurniawan07 commented:
So... anything else I have to do here?

thyageshm commented:
This one is good to go.  Just open a new issue. 
@damithc read for your review

damithc commented:
Apply `toMerge` label to trigger my review.

thyageshm commented:
Sorry sir, I don't have access to my laptop right now and I there are no options to change the labels in the mobile site.
